### ISSUE TITLE
suppert Easyflash save feature

### ISSUE THREAD

**Issue by Flexmaen** [GideonZ/1541ultimate#52]
The Easyflash specification allows to save things on the cartridge.

This is a feature more and more cartridge releases use. Since the 1541 does not support it yet, **saving a game is impossible for recent releases**.
So, it would be great to support the full Easyflash specitications.

For example, some releases that make use of the save feature:
https://csdb.dk/release/?id=156948
https://csdb.dk/release/?id=98674

**Comment by Grrrolf** [comment]
From the [FAQ](https://1541u-documentation.readthedocs.io/en/latest/faq.html#does-the-u2-and-u2-support-easyflash-cartridge-images)

> Does the U2 and U2+ support Easyflash Cartridge images?
> Yes, both carts support EF. Since firmware 3.2 there is also support for saving to the EF cartridge image:
> 
> Technology Preview: EAPI Support for Easyflash, incl. but not limited to writing support. Please note that the changed crt file is NOT saved automatically. You can save it by “F5”, “Save Easyflash crt” in the filebrowser (and optionally renaming the saved file).

So, while running an EF cart, in essence, enter the Ultimate menu, press F5 (contextual menu) and then choose 'Save EasyFalsh crt'. Make sure to rename it afterwards.

While this is might not the type of solution you'd expect, it is the only (feasible) solution right now and it beats having no way of saving.

**Comment by Flexmaen** [comment]
Oh, didn't know that. Sounds good, but i completely froze the Ultimate on my C128 (using Firmware 3.2 (10F)).
Should this work like a cartridge freeze?

**Comment by Flexmaen** [comment]
This also happens on my SX64 and does not seem to be a C128 related problem.
Btw, I have the Ultimate II (without +).

**Comment by markusC64** [comment]
USB Stick or Micro SD?

**Comment by Flexmaen** [comment]
microSD

**Comment by michas0602** [comment]
Agree to the starter. Even though actual way it "beats having no way of saving", saving it correct in the same file would make it 'Ultimate'.

**Comment by Flexmaen** [comment]
But like I said: Even the workaround doesn't work for me yet.

**Comment by marty41** [comment]
Did you wait some time, as saving the EF3 image does take a minute or even longer on the U2. This was tested working on U2 with a 250407 board and 3.2 and 3.3 firmware.

**Comment by Flexmaen** [comment]
> Did you wait some time, as saving the EF3 image does take a minute or even longer on the U2. This was tested working on U2 with a 250407 board and 3.2 and 3.3 firmware.

Ah yes you are right... I did wait a while but expected some kind of dialogue.
Instead there will be a module.crt in your directory. If it is already there, nothing will be saved. So now I understand why to rename the file. Didn't notice it was there, since there were too many files in that directory. Sorry for confusion.

So great, at least there is a workaround.

**Comment by marty41** [comment]
So this works as designed and isn't a workaround and it for sure isn't an issue at all. 
-> Can be closed!

**Comment by markusC64** [comment]
> P.S.: 3.3 isn't for the U2 yet, is it?

Officially not - but if you do not need Ethernet, unofficial 3.3 build will work great. If you need Ethernet, there are non-negligeable chances that Ethernet is not working well. But removing the Ethernet adapter from usb will put the U2 nack to working when using unofficial 3.3 builds.

**Comment by Flexmaen** [comment]
> So this works as designed and isn't a workaround and it for sure isn't an issue at all.
> -> Can be closed!

I don't think that this should be closed. It is not urgent since there is a workaround but maybe it can be marked as enhancement since it makes sense to improve Easyflash support.

**Comment by bugjam64** [comment]
Agree with Flexmaen - there are probably no news on this issue, but I thought I'd push it up anyway.
I remember that you, Gideon, did explain the problem in detail once in a post in the facebook group, but it is impossible to find again - maybe you can find it (you might remember a keyword from the post that allows you to find it back).

**Comment by acarmony1** [comment]
Using 3.7 1541Ultimate as well as my U64. Has the Save Easyflash crt option been removed? SID settings disabled as per FAQ (reu as well) any cart used hitting F5 in file browser mode displays no save easyflash crt (or any option to save crt) ?

**Comment by markusC64** [comment]
The option to save easyflash will only be shown on the f5 menu if

1st) An easyflash module has been loaded.
2nd) Current path in file browser is writable, i. e. in the root where you select the device to save to the option to save is never shown.

Okay, on U64 I have tested 3.8, but it works for me.

**Comment by acarmony1** [comment]
Sorry, I should clarify my U64 is 1.34 (also says v3.8), and my 1541ulitimate 2+ is v3.7. I can create a d64 file, which indicates it is writable (I assume?). I have tried several different EF relase of games (Pirates, Zak Manianc Man and others). After using the save inside the game, pressing freezer button and F5 simply does not show save easyflash option. it has save EDID file, save config to flash and disk and other options. What EF cart did you specifically test it with? I can post a video later if need be? I even went back to V1.28 on U64, still nothing. Reset default settings, turned everything I could off. Still nothing.

**Comment by markusC64** [comment]
Well, I have tried https://csdb.dk/release/?id=107705

Of cause, posting a video (i. e. a link to that video) might help. But we cannot promise that it will help.

**Comment by acarmony1** [comment]
I tried the link you listed. Interestingly, on the U64 when high score is saved at end of event the U64 crashes (locks up with cal games, pops up unreadable ML monitor with winter games). Otherwise displays no save easyflash crt. With 1541U2+ the screen flashes as if saving to easyflash. However still does not display save easyflash crt in menu. I tried the 1541U in 2 different machines. Any suggestions on getting more diagnostic info?

**Comment by Grrrolf** [comment]
acarmony1 wrote:
> However still does not display save easyflash crt in menu. 

Sure it shows the "Save  Easyflash" item in the contextual menu (F5). However, you need to scroll down (press F7 for page down a few times) far enough to see the option.

![image](https://user-images.githubusercontent.com/10052794/93633923-08016d80-f9f0-11ea-8c4d-38471fb020ba.png)

**Comment by acarmony1** [comment]
... Ok .. Found it by scrolling down.. Gonna go sit way back in dark cave now.. Much thanks!

**Comment by Grrrolf** [comment]
It looks like this ticket can be closed now?

**Comment by bugjam64** [comment]
Wellll... Actually it is still only a workaround (and AFAIR it is not even working in all cases). I know that there is no space left in the firmware for any additional code, but I am still hoping for a (branch) version where saving directly to the EF image (and not having to re-save the whole crt) is implemented, at the cost of some other feature which is less used/not essential.

**Comment by GideonZ** [comment]
The patch that is required to make "Save EF" work for all non-compliant CRT
files will be released soon. I have not been able to do this yet due to
personal circumstances.
Re "other features that are not essential" => because you can manually save
the image, making it automatic is therefore not essential by definition.
(There is another way to do it.)
